src/sampler.jl

```diff
@@ -88,125 +87,68 @@ function Adaptation.adapt!(
     return h, κ, isadapted
 end
 
-"""
-Progress meter update with all trajectory stats, iteration number and metric shown.
-"""
-function pm_next!(pm, stat::NamedTuple)
-    ProgressMeter.next!(pm; showvalues=[tuple(s...) for s in pairs(stat)])
+#################################
+### AbstractMCMC.jl interface ###
+#################################
+struct HamiltonianModel{H} <: AbstractMCMC.AbstractModel
+    hamiltonian :: H
 end
 
-"""
-Simple progress meter update without any show values.
-"""
-simple_pm_next!(pm, stat::NamedTuple) = ProgressMeter.next!(pm)
-
-##
-## Sampling functions
-##
+struct HMCState{
+    TTrans<:Transition,
+    TAdapt<:Adaptation.AbstractAdaptor
+}
+    i::Int
+    transition::TTrans
+    adaptor::TAdapt
+end
 
-sample(
-    h::Hamiltonian,
-    κ::AbstractMCMCKernel,
-    θ::AbstractVecOrMat{<:AbstractFloat},
-    n_samples::Int,
+function AbstractMCMC.step(
+    rng::AbstractRNG,
+    model::HamiltonianModel,
+    spl::HMCKernel;
+    init_params, # TODO: implement this. Just do `rand`? Need dimensionality though.
```

Comment:
I can see multiple ways to do this but it's unclear what's the best.
Can you file a PR/discussion to chat on this?

Reply:
Will do! But for now I'll just extract the dimensionality from the metric and use `rand`. But yeah, I'll open an issue after this PR so we can figure out potential better defaults.